include/eve/module/complex/regular/detail/arithmetic.hpp

```diff
@@ -138,4 +138,72 @@ namespace eve::detail
     using z_t = decltype(z1+z2);
     return z_t{eve::average(real(z1), real(z2)), eve::average(imag(z1), imag(z2))};
   }
+
+  //==============================================================================================
+  //  Ternary functions
+  //==============================================================================================
+
+  EVE_FORCEINLINE auto complex_ternary_dispatch( eve::tag::fam_
+                                              , auto const& z1, auto const& z2, auto const& z3
+                                              ) noexcept
+  {
+    return fma(z2, z3, z1);
+  }
+
+  EVE_FORCEINLINE auto complex_ternary_dispatch( eve::tag::fma_
+                                              , auto const& z1, auto const& z2, auto const& z3
+                                              ) noexcept
+  {
+    using z_t = decltype(z1*z2+z3);
+    auto [rz1, iz1] = z_t(z1);
+    auto [rz2, iz2] = z_t(z2);
+    auto [rz3, iz3] = z_t(z3);
+    return z_t{fma(rz1, rz2, fnma(iz1, iz2, rz3)), fma(iz1, rz2, fma(rz1, iz2, iz3))};
+  }
+
+  EVE_FORCEINLINE auto complex_ternary_dispatch( eve::tag::fms_
+                                              , auto const& z1, auto const& z2, auto const& z3
+                                              ) noexcept
+  {
+    using z_t = decltype(z1*z2+z3);
+    auto [rz1, iz1] = z_t(z1);
+    auto [rz2, iz2] = z_t(z2);
+    auto [rz3, iz3] = z_t(z3);
+    return z_t{fma(rz1, rz2, fnms(iz1, iz2, rz3)), fma(iz1, rz2, fms(rz1, iz2, iz3))};
+  }
+
+  EVE_FORCEINLINE auto complex_ternary_dispatch( eve::tag::fsm_
+                                              , auto const& z1, auto const& z2, auto const& z3
+                                              ) noexcept
+  {
+    return fms(z2, z3, z1);
```

Comment:
Isn't it somethign the normal fsm etc does ? Shouldn't we fix fsm_ et al. to accept value instead ?

Reply:
je te propose de mettre ça en attente. Merger comme ça et qd ce sera le tour des concepts on écrémera